### DISCUSSION TITLE
New deno mirror on npmjs.org?

### DISCUSSION THREAD

**Discussion by bjesuiter** [denoland/deno#15752] [Ideas]
Hi there, 

I would like to have an official deno binary mirror on npmjs.org.

## TL;DR Questions 

@ry @lucacasonato 

(Target for these questions is mostly the deno company) 

1. Do you want to build an official deno binary mirror package on npmjs.org? 

2. If not, would you allow me to start/do it? (See my use-case and my plan below)

3. Would you be down to let me contribute my solution to an official deno namespace on npmjs? (If you have any and clearly after review from the deno company) 

## Use-Case

I work at a company behind a firewall (as many people).

I would like to use deno as a lightweight scripting environment to avoid long-running npm installs in ci. 

I also would like to use deno'a fresh package for lightweight status pages about our running services. 

I'm able to install deno on my Developer machine, since it doesn't need admin rights. 
However, I can't install it in CI/CD environment, because the Jenkins server has no real internet connection. 

_A short description of our Jenkins environment: 
We have an openshift kubernetes cluster, which contains the Jenkins server. Each build node for this Jenkins is an ephemeral container which gets started for each individual build und then discarded.
Note: I can't add a new base image for build nodes, due to permissions._

The only solution we have to get dependencies on the Jenkins build nodes is a company-wide nexus npm proxy registry. 

## What I tried before 

I tried to install deno via official methods on the Jenkins build runners. This has several drawbacks: 

1. I have to provide the deno executable via my projects git repo, since I can't load it from the internet. 
(Which is a huuuuge pain in the ass already, due to binaries in the git repo) 

2. The installation does not persist due to these ephemeral build nodes. This increases complexity of build scripts (due to not having deno in the main PATH) and all 'global' scripts I would install with deno would have to be installed with every build run.

## My Proposed Solution
_This is the proposal for the question in TL;DR Section above, how I would do it!_

1. I would like to set up a git repo, which gets notified of each new deno binary release. 
Options: 
- maybe there is a webhook somewhere for that? 
- or I have to Setup a small deno.deploy service, which regularly checks the version of the deno executable at the normal download endpoint. 

After that: 
2. When a new deno binary is out, this event should trigger a GitHub actions pipeline, which downloads the new binaries and packages them into a node package. 

One open question for that is: How to structure the npm package best? 

- Can we add different os  binaries in the package.json bin fields for npm to pick the right one to execute based on host system? 

- Or do we have to write a little js loader in node to know which executable to run based on the host os? 

- could we execute this os-detection earlier (for example in the preinstall hook) to not having to launch deno through nodejs every time?

- if we cannot do that, do we release one deno package per system and arch and add that to the naming of the package?

3. Lastly these packages simply need to be published on npmjs.org

I'm very interested in the answers!

**Comment by katakatana** [comment]
Benjamin,
I have no idea 🤷‍♂️
I have to go but know that I’ve reached the end.

**Reply by bjesuiter** [reply]
Maybe I can boil this question down a little: 

Am I allowed to redistribute deno binaries to npmjs.org via automated GitHub actions?

**Reply by bjesuiter** [reply]
Bzw. Thanks for reading! 😀🥰

**Comment by AnInternetTroll** [comment]
Have you looked into [deno-bin](https://www.npmjs.com/package/deno-bin)?

**Reply by bjesuiter** [reply]
Oh nice! I must've overlooked this! Thank you for bringing this up! 🤩

**Reply by bjesuiter** [reply]
I have to look more into it though... 

The problem is, that my CI/CD environment does not have real internet and it may not be able to download deno via this package. 
But I will check this, probably next week, when I'm back at work from vacation 😄
